### ISSUE TITLE
Add compatibility with Github's GraphQL API v4 and rest API v3

### ISSUE THREAD

**Issue by marusinm** [packit-service/ogr#79]
I'm trying to implement ogr into [release-bot](https://github.com/user-cont/release-bot) and it's a little bit painful without documentation from ogr. 

I was playing with ogr on my Github projects and everything is working fine. When I tried to reimplement [release-bot's](https://github.com/user-cont/release-bot) functionalities which communicate with Github via GraphQL or rest API I got stuck.  

One of the examples could be comments. I want to implement commenting on Github issues and PRs via ogr. 
- In [release-bot](https://github.com/user-cont/release-bot) it works via graphQL and sending some `node_id` inside query which recognizes a PR or an Issue. 
- In ogr there is `pr_comment` function in ogr/services/github.py. If am I right this is an alternative I'm looking for. However, I don't fully understand the function's parameters since it needs something else then API query.

I understand that this could be a very complex problem since ogr need to preserve compatibility with all git forges. I'm just looking for some starting point on how to change GraphQL/rest API with ogr. Maybe I just misunderstand some ogr's fundamentals.

**Comment by lachmanfrantisek** [comment]
Yes, you are looking at the right function:

```python
    def pr_comment(
        self,
        pr_id: int,
        body: str,
        commit: str = None,
        filename: str = None,
        row: int = None,
) -> "PRComment":
```

When creating a comment, you probably need an ID of the pull-request and text to send (=`body`). (Other parameters allows you to comment on code, etc.. You can ignore them for your purpose.)

But you need to call it from the `GithubProject` class:

1. create an instance of `GithubService` (with credentials)
2. get an instance of the project (`service.get_project`)
3. send a comment (`project.pr_comment(pr_id=3, body="Helloworld!")`)

The purpose of OGR is to have one API for multiple services, so we cannot match the naming of all forges...

**Comment by marusinm** [comment]
Yes, I completely understand what you answered me. 

> But you need to call it from the GithubProject class:
> 1. create an instance of GithubService (with credentials)
> 2. get an instance of the project (service.get_project)
> 3. send a comment (project.pr_comment(pr_id=3, body="Helloworld!"))

I also did these steps before. And I understand `pr_comment` function's definition. 

When we look at our model situation with comments this confuses me:
1. To get all PRs I call `prs = project.get_pr_list(PRStatus.all)` to get all PRs
2. Then I try to get first PR's id and comment it with
 `project.pr_comment(pr_id=prs[0].id, body="release-bot comment test")`
3. I obtain this error. It seems that the integer id which I pass to `pr_comment` function isn't correct.
```Traceback (most recent call last):
  File "/Users/Marek/Documents/Python/ogr/experiments.py", line 17, in <module>
    project.pr_comment(prs[0].id, "release-bot comment test")
  File "/Users/Marek/Documents/Python/ogr/ogr/mock_core.py", line 49, in readonly_func
    return func(self, *args, **kwargs)
  File "/Users/Marek/Documents/Python/ogr/ogr/services/github.py", line 225, in pr_comment
    github_pr = self.github_repo.get_pull(number=pr_id)
  File "/Users/Marek/Documents/Python/ogr/venv/lib/python3.6/site-packages/github/Repository.py", line 2236, in get_pull
    self.url + "/pulls/" + str(number)
  File "/Users/Marek/Documents/Python/ogr/venv/lib/python3.6/site-packages/github/Requester.py", line 275, in requestJsonAndCheck
    return self.__check(*self.requestJson(verb, url, parameters, headers, input, self.__customConnection(url)))
  File "/Users/Marek/Documents/Python/ogr/venv/lib/python3.6/site-packages/github/Requester.py", line 286, in __check
    raise self.__createException(status, responseHeaders, output)
github.GithubException.UnknownObjectException: 404 {'message': 'Not Found', 'documentation_url': 'https://developer.github.com/v3/pulls/#get-a-single-pull-request'}

**Comment by lachmanfrantisek** [comment]
Yes, your approach seems good.

> I obtain this error. It seems that the integer id which I pass to pr_comment function isn't correct.

Please, check the received objects and the id you are passing, so we can see, which one is causing the problem. Does the pull-request with that number exist? (Are you sure you are not working with the fork?)

**Comment by marusinm** [comment]
>Please, check the received objects and the id you are passing, so we can see, which one is causing the problem. Does the pull-request with that number exist? (Are you sure you are not working with the fork?)

I already checked it. I am definitely don't work with fork. I'm passing Integer id which is saved in class `PullRequest.id` within abstract.py. Such instances are created after calling `project.get_pr_list(PRStatus.all)`.  Everything seems right. Pull-request really exists but it is already merged, could be this problem?

Example of code which returns an error from the previous comment:
```Python
from ogr.services.github import GithubService
from ogr.abstract import PRStatus

service = GithubService(token="xyz")
project = service.get_project(
    repo="repo-name",
    namespace="username"
)
prs = project.get_pr_list(PRStatus.all)
#prs[0].id => this is existing PR which return integer id
project.pr_comment(pr_id=prs[0].id, body="Hello world from ogr")

**Comment by lachmanfrantisek** [comment]
> Everything seems right. 

Yes, everything looks good...

---

Please:
- Check the data in `prs`. Are the other attributes right?
- Check this url by hand: `https://github.com/{my-org}/{my-project}/pulls/{pr-id}`
- And this one: `https://api.github.com/{my-org}/{my-project}/pulls/{pr-id}`

**Comment by lachmanfrantisek** [comment]
> Pull-request really exists but it is already merged, could be this problem?

Good point. It does not work for merged PRs...

**Comment by marusinm** [comment]
Now, I see where is the problem. 

This is example of one of the pull-requests obtaned with `project.get_pr_list(PRStatus.all)`:
```
PullRequest(title='0.2.4 release', id=289221055, status='open', url='https://github.com/{org}/{repo}/pull/38', description='Hi,
 you h...', author='Marek Marušin', source_branch='0.2.4-release', target_branch='master', created='2019-06-18 11:10:25')
```

What I was doing is that I was working with `id` of this pull-request in this case `id=289221055`. But I have to work with id from `url` so `url='https://github.com/{org}/{repo}/pull/38'` in this case number 38. This was confusing for me. Now it is working. The question is why these ids are different. I need to use one from `url` but now I have to parse it. 

Actually, this is the reason why I opened this issue. I would expect that id will be the same as uses graphQL which is named `node_id` and it is kind of hash. 

>Good point. It looks like it does not work for merged PRs...

This isn't true. I tried it on opened and also merged PR and it is working when I use correct id.

**Comment by lachmanfrantisek** [comment]
>  The question is why these ids are different. 

OK, that's definitely a problem. This looks like some parsing problem or it can be a "global-pr-id". It should be fixed to be the same.

> This isn't true. I tried it on opened and also merged PR and it is working when I use correct id.

Maybe some problem on my side when I've played with that.

**Comment by marusinm** [comment]
>OK, that's definitely a problem. This looks like some parsing problem or it can be a "global-pr-id". It should be fixed to be the same.

It indeed looks like some "global-pr-id" downloaded via some github lib. Can I work on it? If yes, should I just parse `url` attribute or do you think there is a better way?

I would like to ask if ogr has also some special function for commenting on Github issues or If I can use `pr_comment` again.

**Comment by lachmanfrantisek** [comment]
> Can I work on it?

Would be nice! Thanks.

We can use `number` instead of `id` here:
https://github.com/packit-service/ogr/blob/master/ogr/services/github.py#L316

 ([gitpython docs](https://pygithub.readthedocs.io/en/latest/github_objects/Project.html))


> I would like to ask if ogr has also some special function for commenting on Github issues or If I can use pr_comment again.

On the first look, we have no `pr_comment`. It's a shame, but should be (very) easy to implement.

**Comment by lachmanfrantisek** [comment]
I am going to close this one since it was partially solved in #81 and covered by #80.

@marusinm Feel free to open other specific issues for problems that you found...
